### PR TITLE
Implement support for `rack.protocol`.

### PR DESCRIPTION
**Pull request by ioquatix**
### Description

Implement support for `rack.protocol` which is an optional feature of Rack 3.1, required to support `protocol-websocket`.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added (or updated) appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by dentarg** [comment]
Do you mind providing some more background on this? For anyone that hasn't followed along with the latest Rack developments. Any links to discussions/issues are appreciated too.

**Comment by ioquatix** [comment]
It has been discussed here: https://github.com/rack/rack/issues/1946 and there are linked issues/discussions which provide more context.

File: lib/puma/request.rb [hunk 1]
```diff
@@ -104,6 +108,12 @@ def handle_request(client, requests)
           status, headers, app_body = [501, {}, ["#{env[REQUEST_METHOD]} method is not supported"]]
         end
 
+        if protocol = headers.delete(RACK_PROTOCOL)
+          status = 101
+          headers['upgrade'] = protocol
+          headers['connection'] = 'upgrade'
```
Comment (ioquatix) [inline lib/puma/request.rb:114]: We could make these string constants if preferred.

File: lib/puma/request.rb [hunk 2]
```diff
@@ -104,6 +108,12 @@ def handle_request(client, requests)
           status, headers, app_body = [501, {}, ["#{env[REQUEST_METHOD]} method is not supported"]]
         end
 
+        if protocol = headers.delete(RACK_PROTOCOL)
+          status = 101
```
Comment (ioquatix) [inline lib/puma/request.rb:112]: I think it's advisable we force the status code here. HTTP/2 uses `200 Ok` but HTTP/1 uses 101... There is no reason to use anything other than 101 AFAIK.

**Comment by ioquatix** [comment]
Unrelated failure:

```
TestPumaServer#test_chunked_request_invalid_extension_header_length = 0.00 s = .
(libev) kqueue kevent: Bad file descriptor
/Users/runner/work/_temp/139c090b-de4a-473c-b06a-14f67de45c34.sh: line 1:  6934 Abort trap: 6           test/runner --verbose
TestPumaServer#test_chunked_encoding = 
```

That seems... odd?

**Comment by ioquatix** [comment]
Rack 3.1 was released with `rack.protocol` in the SPEC. It would be great to merge this PR.

**Comment by ioquatix** [comment]
Actually, after consideration, I updated `protocol-rack` to support `HTTP_UPGRADE` as a fallback - if `rack.protocol` is not present - and it works okay for Puma. So this is completely optional.

**Pull request closed without merging**